[UI] Add Toc in Sidebar

senwang86 opened the pull request:
## Summary
- Toc enables the viewport center.
- Will add more follow-up UI tuning in future

## Test
`yarn add @mui/lab @mui/material`


![toc](https://github.com/codepod-io/codepod/assets/10226241/b36a2408-348d-46bc-a07b-8656517f08ff)

lihebi commented on ui/src/components/Sidebar.tsx:
```diff
@@ -1168,6 +1178,123 @@ function ExportButtons() {
+  );
+}
+
+function dfs(
+  nodeId: string,
+  toc: Map<string, ReactflowNode[]>,
+  nodesMap: Y.Map<ReactflowNode<NodeData>>
+) {
+  let jsx = <></>;
+
+  if (toc.has(nodeId) && toc.get(nodeId)!.length > 0) {
+    toc.get(nodeId)?.forEach((child) => {
+      jsx = (
+        <>
+          {jsx}
+          <TocPodButton id={child.id} />
+          {toc.has(child.id) ? <ul>{dfs(child.id, toc, nodesMap)}</ul> : null}
+        </>
+      );
+    });
+  }
+
+  return jsx;
+}
+
```
This isn't "React style" code.

```ts
// Component must be CamelCase
function ToC({id}) {
   if (!toc.has(id)) return null; // Return early
   const children = toc.get(id)
   return ( // directly return the Component, not assign and return a variable
     <Box>
        <TocPodButton id={child.id} />
        {children.map(child=>(
           // ToC is used directly as a component name.
           // A non-CamelCase function cannot be used as a component name.
            <ToC id={child}/> 
         ))}
      </Box>)
}
```

lihebi commented on ui/src/components/Sidebar.tsx:
```diff
@@ -1168,6 +1178,123 @@ function ExportButtons() {
+        <>
+          {jsx}
+          <TocPodButton id={child.id} />
+          {toc.has(child.id) ? <ul>{dfs(child.id, toc, nodesMap)}</ul> : null}
```
You are using `<ul>` to control padding and using `ListItem` to show lists. These two things sound like two alternatives to achieve the same functionality, not to be used together.

How about using MUI's [TreeView](https://mui.com/material-ui/react-tree-view/)? It seems to offer a good padded UI with additional features such as multi-selection.

senwang86 replied:
[TreeView](https://mui.com/material-ui/react-tree-view/) is indeed a better option.

lihebi commented:
Instead of showing (full) IDs, some options (ordered by priority, high to low):

1. show pod/scope title (.name)
2. show the title/first-line of rich text content
3. show the function defined in code pod (I'll populate the parse result somewhere)
4. show a shortened ID (e.g., first 8 characters)

senwang86 commented:
> Instead of showing (full) IDs, some options (ordered by priority, high to low):
> 
> 1. show pod/scope title (.name)
> 2. show the title/first-line of rich text content
> 3. show the function defined in code pod (I'll populate the parse result somewhere)
> 4. show a shortened ID (e.g., first 8 characters)

[1d71993](https://github.com/codepod-io/codepod/pull/458/commits/1d719936045b53804c2a99f10b8039f628290213) addresses comments.

senwang86 commented:
> > 4. show a shortened ID (e.g., first 8 characters)
> 
> [1d71993](https://github.com/codepod-io/codepod/pull/458/commits/1d719936045b53804c2a99f10b8039f628290213) addresses comments.

lihebi commented:
Nice, thanks!
